src/shared/IO/Networking/AsyncSocket.h

```diff
@@ -78,45 +121,61 @@ namespace IO { namespace Networking {
             std::size_t m_writeSrcBufferBytesLeft = 0;
 #endif
     };
-
-    template<typename SocketType>
-    AsyncSocket<SocketType>::~AsyncSocket() noexcept(false)
-    {
-        sLog.Out(LOG_NETWORK, LOG_LVL_DETAIL, "Destructor called ~AsyncSocket: No references left");
-        if (!m_disconnectRequest)
-            CloseSocket();
-
-        MANGOS_ASSERT(!HasPendingTransfers());
-    }
-
-    template<typename SocketType>
-    IO::Networking::IpEndpoint const& AsyncSocket<SocketType>::GetRemoteEndpoint() const
-    {
-        return m_socket.m_peerEndpoint;
-    }
-
-    template<typename SocketType>
-    std::string AsyncSocket<SocketType>::GetRemoteIpString() const
-    {
-        return GetRemoteEndpoint().ip.toString();
-    }
-
-    template<typename SocketType>
-    void AsyncSocket<SocketType>::ReadSkip(std::size_t skipSize, std::function<void(IO::NetworkError const&)> const& callback)
-    {
-        std::shared_ptr<std::vector<uint8_t>> skipBuffer(new std::vector<uint8_t>());
-        skipBuffer->resize(skipSize);
-        Read((char*)skipBuffer->data(), skipSize, [skipBuffer, callback](IO::NetworkError const& error)
-        {
-            // KEEP skipBuffer in scope!
-            // Do not remove skipBuffer before Read() is done, since we are transferring into it via async IO
-            // and since we are using a raw pointer, the Task has no knowledge about the lifetime of the std::vector
-            skipBuffer->clear();
-            callback(error);
-        });
-    }
 }} // namespace IO::Networking
 
+template<typename SocketType>
+IO::Networking::AsyncSocket<SocketType>::~AsyncSocket() noexcept(false)
+{
+    sLog.Out(LOG_NETWORK, LOG_LVL_DETAIL, "Destructor called ~AsyncSocket: No references left");
+    CloseSocket();
+
+//#ifdef DEBUG
+    // Logic behind these checks:
+    // If the destructor is called, there should be no more std::shared_ptr<> references to this object
+    // Every Read(...) or Write(...) should use `shared_from_this()` if this is not the case, one of the following checks will fail
+    int state = m_atomicState.load(std::memory_order::memory_order_relaxed);
+    if (state & SocketStateFlags::CONTEXT_PRESENT)
+            MANGOS_ASSERT(false); // TODO: allow MANGOS_ASSERT to accept a string description
```

Comment:
simply remove the if statement before this and change it to
MANGOS_ASSERT(!(state & SocketStateFlags::CONTEXT_PRESENT))

Reply:
lol. I think I used this part for something else first and added the `MANGOS_ASSERT(false)` later.
But you are right, inlining the condition is the correct way.